### PR TITLE
add ZManaged#scopeIO

### PR DESCRIPTION
**Pull request by mschuwalow**
I don't know whether it's worth having as a combinator, but I it found very useful to have.

/cc @iravid

### CODE REVIEW

File: core/shared/src/main/scala/zio/ZManaged.scala [hunk 1]
```diff
@@ -1539,12 +1543,26 @@ object ZManaged {
...
+    }
+
+  /**
+   * Creates a scope in which resources can be safely allocated into together with a release action.
+   */
+  def scopeIO[R]: ZManaged[R, Nothing, ScopeIO] =
     ZManaged {
       // we abuse the fact that Function1 will use reference equality
-      Ref.make(Set.empty[Exit[Any, Any] => ZIO[R, Nothing, Any]]).map { finalizers =>
+      Ref.make(Set.empty[Exit[Any, Any] => ZIO[Any, Nothing, Any]]).map { finalizers =>
```
Comment (iravid) [inline core/shared/src/main/scala/zio/ZManaged.scala]:
```suggestion
      Ref.make(Set.empty[Exit[Any, Any] => UIO[Any]]).map { finalizers =>
```

File: core/shared/src/main/scala/zio/ZManaged.scala [hunk 2]
```diff
@@ -1539,12 +1543,26 @@ object ZManaged {
    * Creates a scope in which resources can be safely preallocated.
    */
   def scope[R]: ZManaged[R, Nothing, Scope] =
```
Comment (iravid) [inline core/shared/src/main/scala/zio/ZManaged.scala]: How about we rename `scope` to `preallocationScope`, and give the `scope` names to the new combinator?

Comment (iravid) [reply core/shared/src/main/scala/zio/ZManaged.scala]: This way I wouldn't be entirely lying here: https://twitter.com/iravid_/status/1180569798037708800?s=20 😇

Comment (mschuwalow) [reply core/shared/src/main/scala/zio/ZManaged.scala]: Nice, will update 😁

File: core/shared/src/main/scala/zio/ZManaged.scala [hunk 3]
```diff
@@ -1665,6 +1648,41 @@ object ZManaged {
   def sandbox[R, E, A](v: ZManaged[R, E, A]): ZManaged[R, Cause[E], A] =
     v.sandbox
 
+  /**
+   * Creates a scope in which resources can be safely allocated into together with a release action.
+   */
+  def scope[R]: ZManaged[R, Nothing, Scope] =
```
Comment (iravid) [inline core/shared/src/main/scala/zio/ZManaged.scala]:
```suggestion
  val scope: ZManaged[R, Nothing, Scope] =
```

File: core/shared/src/main/scala/zio/ZManaged.scala [hunk 4]
```diff
@@ -1538,35 +1542,14 @@ object ZManaged {
...
-            }
-          },
-          release = exitU =>
-            for {
-              fs    <- finalizers.get
-              exits <- ZIO.foreachPar(fs)(_(exitU).run)
-              _     <- ZIO.done(Exit.collectAllPar(exits).getOrElse(Exit.unit))
-            } yield ()
-        )
+  def preallocationScope[R]: ZManaged[R, Nothing, PreallocationScope] =
```
Comment (iravid) [inline core/shared/src/main/scala/zio/ZManaged.scala]:
```suggestion
  val preallocationScope: ZManaged[Any, Nothing, PreallocationScope] =
```

**Comment by mschuwalow** [comment]
Ah sorry, didn't pay attention with the type parameters.
Scope will need to stay a def though or we have to break our alphabetic ordering

**Comment by iravid** [comment]
Ah sure. We can just drop the outer `R` on it then.

**Pull request merged**